### ISSUE TITLE
Ingest JSON processor parses starting object instead of failing

### ISSUE THREAD

**Issue by phillbaker** [elastic/elasticsearch#43262]
**Elasticsearch version** (`bin/elasticsearch --version`):
6.2.3

**Plugins installed**: 
analysis-icu        
analysis-kuromoji   
analysis-phonetic   
analysis-seunjeon   
analysis-smartcn    
analysis-stempel    
analysis-ukrainian  
elasticsearch-jetty 
ingest-attachment   
ingest-user-agent   
mapper-murmur3      
mapper-size         

**JVM version** (`java -version`):
OpenJDK 64-Bit Server VM 12.0.1

**OS version** (`uname -a` if on a Unix-like system):
CentOS Linux 7 (Core)

**Description of the problem including expected versus actual behavior**:

The ingest JSON processor treats a value that starts with a valid JSON object or array as entirely that JSON object instead of failing on the non-JSON values following the object. This is very unexpected as most (all?) other json converters will fail on these inputs.

To summarize:
* `[360113.865822] foo` (string) => `[360113.865822]` (object), expected failure
* `{"foo": "bar"} baz` (string) => `{"foo": "bar"}` (object), expected failure

**Steps to reproduce**:

```
POST /_ingest/pipeline/_simulate
{
  "pipeline": {
    "processors": [{
      "json": {
        "field": "message",
        "add_to_root": false,
        "target_field": "fields",
        "ignore_failure" : true
      }
    }
  ]
  }, 
  "docs" : [
    { "_source":{"message": "[360113.865822] wbrdg-0afe001ce"}},
    { "_source":{"message": "foo [360113.865822] wbrdg-0afe001ce"}},
    { "_source":{"message": "{\"foo\":\"bar\"} wbrdg-0afe00e"}}
  ]
}
```

Result:
```
{
  "docs": [
    {
      "doc": {
        "_index": "_index",
        "_type": "_type",
        "_id": "_id",
        "_source": {
          "message": "[360113.865822] wbrdg-0afe001ce",
          "fields": [
            360113.865822
          ]
        },
        "_ingest": {
          "timestamp": "2019-06-17T01:53:35.442134Z"
        }
      }
    },
    {
      "doc": {
        "_index": "_index",
        "_type": "_type",
        "_id": "_id",
        "_source": {
          "message": "foo [360113.865822] wbrdg-0afe001ce"
        },
        "_ingest": {
          "timestamp": "2019-06-17T01:53:35.44217Z"
        }
      }
    },
    {
      "doc": {
        "_index": "_index",
        "_type": "_type",
        "_id": "_id",
        "_source": {
          "message": "{\"foo\":\"bar\"} wbrdg-0afe00e",
          "fields": {
            "foo": "bar"
          }
        },
        "_ingest": {
          "timestamp": "2019-06-17T01:53:35.442189Z"
        }
      }
    }
  ]
}
```

**Provide logs (if relevant)**:
N/a.

cc @talevy since it seems like you've been pretty involved in the processors work

**Comment by talevy** [comment]
thank you for the ping @phillbaker, The team will review solutions to this when 
they get the chance. I understand that many JSON parsers would fail, but this 
behavior can be viewed as a feature (not a bug) for some. Due to this, if we 
decide to change the behavior, then the change may have to wait for the next
major release.

**Comment by phillbaker** [comment]
Thanks @talevy. I understand how this could be viewed as a feature and how it might be a considered a breaking change. What would you think about a PR to docs that documents this behavior?

As one alternative, what would the team think about adding a flag to the processor to exclude this behavior - perhaps named "strict" where the default is false? For example:
```
{
  "pipeline": {
    "processors": [{
      "json": {
        "field": "message",
        "strict": true,
        "target_field": "fields",
        "ignore_failure" : true
      }
    }
  ]
}
```

**Comment by selfuryon** [comment]
Hello! I have the same issue, but case is worse:
* `aaa 1111` -> wrong parse, expected
* `1111 aaaa` -> parse as `1111`, unexpected

**Steps to reproduce**:

Reqeust:
```
$ curl -X POST "http://localhost:9200/_ingest/pipeline/_simulate?pretty" -H 'Content-Type: application/json' -d'
  {
    "pipeline": {
      "processors": [{
        "json": {
          "field": "message",
          "target_field": "extended",
          "ignore_failure" : true
        }
      }
    ]
    },
    "docs" : [
      { "_source":{"message": "aaaaa 11111"}},
      { "_source":{"message": "11111 aaaaa"}}
    ]
  }
  '
```
Response:
```
{
  "docs" : [
    {
      "doc" : {
        "_index" : "_index",
        "_type" : "_doc",
        "_id" : "_id",
        "_source" : {
          "message" : "aaaaa 11111"
        },
        "_ingest" : {
          "timestamp" : "2020-04-01T12:45:17.472722Z"
        }
      }
    },
    {
      "doc" : {
        "_index" : "_index",
        "_type" : "_doc",
        "_id" : "_id",
        "_source" : {
          "message" : "11111 aaaaa",
          "extended" : 11111
        },
        "_ingest" : {
          "timestamp" : "2020-04-01T12:45:17.472727Z"
        }
      }
    }
  ]
}
```

Why it is worse? As Example, if you have multiline outputs and some of them can start from numbers. 
Especially, in kubernetes, where some logs can be in json but others not. Example, cp-kafka with some statistics:
```
[2020-04-01 13:08:58,666] INFO [kafka-log-cleaner-thread-0]:
	Log cleaner thread 0 cleaned log __consumer_offsets-1
	0.0 MB of log processed in 0.1 seconds (0.0 MB/sec).
	Indexed 0.0 MB in 0.1 seconds (0.0 Mb/sec, 69.4% of total time)
	Buffer utilization: 0.0%
	Cleaned 0.0 MB in 0.0 seconds (0.1 Mb/sec, 30.6% of total time)
	Start size: 0.0 MB (13 messages)
	End size: 0.0 MB (6 messages)
	27.6% size reduction (53.8% fewer messages)
 (kafka.log.LogCleaner)
```

**Comment by lucabelluccini** [comment]
I don't know if this issue covers those cases (as above):
```
POST /_ingest/pipeline/_simulate
{
  "pipeline": {
    "description": "",
    "processors" : [
      {
        "json" : {
          "field" : "message"
        }
      }
    ]
  },
  "docs": [
    {
      "_source": {
        "message": """ 1268 : TimeOut = 123 : a"""
      }
    }
  ]
}
```

Output:
```
{
  "docs" : [
    {
      "doc" : {
        "_index" : "_index",
        "_type" : "_doc",
        "_id" : "_id",
        "_source" : {
          "message" : 1268
        },
        "_ingest" : {
          "timestamp" : "2021-04-15T16:39:14.670010703Z"
        }
      }
    }
  ]
}
```

**Comment by zez3** [comment]
Would this merged PR be something related, that might also help solve the issue described here?
https://github.com/elastic/beats/pull/32010

**Comment by zez3** [comment]
and this https://github.com/elastic/beats/issues/31712

**Comment by mattc58** [comment]
We're going to address this issue in a near-term release. This is a bug and the safer course of action for most use cases will be to have strict parsing, and error out on some of the invalid samples provided here. However, we may incorporate a `strict` or `lenient` type of boolean option to allow the user to opt in to the current default (and broken) behavior. The new default option though will be to strictly parse the input and error out as expected.

**Comment by masseyke** [comment]
Closing because this is fixed by #93179
